src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -2875,184 +3030,650 @@ private void ensurePublishCapability(
+                else if (newCapabilities != capabilities)
+                {
+                    this.capabilities = newCapabilities;
+                    doApplicationFlush(traceId, authorization, 0, flags);
+                }
+            }
+
+            private void doApplicationBegin(
+                long traceId,
+                long authorization,
+                long affinity,
+                String topicFilter,
+                int flags,
+                int subscriptionId)
+            {
+                assert state == 0;
+                state = MqttState.openingInitial(state);
+
+                final MqttBeginExFW beginEx = mqttBeginExRW.wrap(extBuffer, 0, extBuffer.capacity())
+                                                           .typeId(mqttTypeId)
+                                                           .capabilities(r -> r.set(valueOf(capabilities)))
+                                                           .clientId(clientId)
+                                                           .topic(topicFilter)
+                                                           .flags(flags)
+                                                           .subscriptionId(subscriptionId)
+                                                           .build();
+
+                router.setThrottle(initialId, this::onApplicationInitial);
+                correlations.put(replyId, this::onApplicationReply);
+                doBegin(application, routeId, initialId, traceId, authorization, affinity, beginEx);
+
+                // final int topicKey = topicKey(topicFilter);
+                // final MutableInteger activeStreams = activeStreamsByTopic.computeIfAbsent(topicKey, key -> new MutableInteger());
+                // activeStreams.value++;
+
+                // if (hasPublishCapability(capabilities))
+                // {
+                //     doSignalPublishExpirationIfNecessary();
+                // }
+            }
+
+            private void doApplicationData(
+                long traceId,
+                long authorization,
+                int reserved,
+                SessionMessageType messageType,
+                OctetsFW payload,
+                Flyweight extension)
+            {
+                assert MqttState.initialOpening(state);
+
+                assert hasPublishCapability(this.capabilities);
+
+                final DirectBuffer buffer = payload.buffer();
+                final int offset = payload.offset();
+                final int limit = payload.limit();
+                final int length = limit - offset;
+                assert reserved >= length + initialPadding;
+
+                initialBudget -= reserved;
+
+                assert initialBudget >= 0;
+
+                sessionStateRO.decodeSessionState(messageType, payload, extension);
+
+                doData(application, routeId, initialId, traceId, authorization, budgetId,
+                    reserved, buffer, offset, length, extension);
```

Comment:
We probably don't need `sessionStateRO` here, and can instead model echo logic based on stream being fully closed or not.
```
if (MqttState.closed(state))
{
    onSessionStateUpdated(payload, extension);
}
else
{
    doData(application, routeId, initialId, traceId, authorization, budgetId,
                    reserved, buffer, offset, length, extension);
}
```